nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java

```diff
@@ -45,25 +52,48 @@
     @Override
     public DocumentModel createCompoundDocument(CoreSession session, String parent, Blob archive) {
         String compoundDocName = FilenameUtils.removeExtension(archive.getFilename());
-        DocumentModel compoundDoc = session.createDocumentModel(parent, compoundDocName, COMPOUND_DOCTYPE);
-        compoundDoc.setPropertyValue("dc:title", compoundDocName);
-        compoundDoc = session.createDocument(compoundDoc);
-        createCompoundDocument(compoundDoc, archive);
-        return compoundDoc;
+        var parentDoc = session.getDocument(new PathRef(parent));
+        try (var zip = new ZipFile(archive.getFile())) {
+            Set<String> entries = zip.stream().map(ZipEntry::getName).collect(Collectors.toSet());
+            String compoundType = runScript(parentDoc, COMPOUND_DOCTYPE_DETECTION_OPERATION, Map.of("entries", entries));
+            DocumentModel compoundDoc = session.createDocumentModel(parent, compoundDocName, compoundType);
+            compoundDoc.setPropertyValue("dc:title", compoundDocName);
+            compoundDoc = session.createDocument(compoundDoc);
+            createCompoundDocument(compoundDoc, zip);
+            return compoundDoc;
+        } catch (OperationException e) {
+            var message = String.format("Could not determine the CompoundDocument type for archive: %s.",
+                    archive.getFilename());
+            throw new NuxeoException(message, e);
+        } catch (IOException e) {
+            var message = String.format("Failed to create CompoundDocument: %s. Something went wrong with archive: %s.",
+                    compoundDocName, archive.getFilename());
+            throw new NuxeoException(message, e);
+        }
     }
 
-    protected void createCompoundDocument(DocumentModel compoundDoc, Blob archive) {
-        try (ZipFile zip = new ZipFile(archive.getFile())) {
-            zip.stream()
-               .filter(this::isAllowedEntry)
-               .sorted(Comparator.comparing(ZipEntry::getName))
-               .forEach(entry -> createEntry(compoundDoc, archive.getFilename(), zip, entry));
-        } catch (IOException e) {
-            String message = String.format(
-                    "Failed to create CompoundDocument: %s. something went wrong with archive: %s.",
-                    compoundDoc.getName(), archive.getFilename());
+    protected String runScript(DocumentModel doc, String scriptId, Map<String, ?> params)
```

Comment:
```suggestion
    protected String runScript(String scriptId, DocumentModel doc, Map<String, ?> params)
```
Seems more readable to read "operation id, input, params".